contracts/liquidity_pool.cairo

```diff
@@ -635,6 +661,124 @@ func _get_value_of_pool_position{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*,
+    let (res: UserPoolInfo*) = alloc();
+    let (upoolinfo_len: felt) = map_and_filter_address_to_userpoolinfo(user, lptoken_addrs, lptoken_addrs_len, res);
+    return (upoolinfo_len, res);
+}
+
+func map_and_filter_address_to_userpoolinfo{
+    syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr
+} (user_addr: Address, lptoken_addrs: Address*, lptoken_addrs_len: felt, upoolinfo: UserPoolInfo*)
+    -> (upoolinfo_len: felt){
+    if (lptoken_addrs_len == 0){
+        return (0,);
+    }
+    let val = get_userpoolinfo(user_addr, lptoken_addrs[0]);
+    if (val.value.low == 0){
+        return map_and_filter_address_to_userpoolinfo(user_addr, lptoken_addrs + 1, lptoken_addrs_len - 1, upoolinfo);
+    }
+    assert [upoolinfo] = val;
+    let (retval: felt) = map_and_filter_address_to_userpoolinfo(user_addr, lptoken_addrs + 1, lptoken_addrs_len - 1, upoolinfo + UserPoolInfo.SIZE);
+    let retval_adj = retval + 1;
+    return (retval_adj,);
+}
+
+// func get_value_of_capital_in_lpool{
+//     syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr
+// } () -> (
+// ) {
+
+// }
+
+@view
+func get_all_poolinfo{
+    syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr
+} () -> (
+    poolinfo_len: felt,
+    poolinfo: PoolInfo*
+) {
+    alloc_locals;
+    let (lptoken_addrs_len: felt, lptoken_addrs: Address*) = get_all_lptoken_addresses();
+    // map address->poolinfo via get_poolinfo
+    let (res: PoolInfo*) = alloc();
+    map_address_to_poolinfo(lptoken_addrs, res, lptoken_addrs_len, 0);
+    //let (array : PoolInfo*) = alloc();
+    //let pool = Pool(0, 0, 0);
+    //let poolinfo = PoolInfo(pool, 0, 0, 0, 0);
+    //assert array[0] = poolinfo;
+    return (lptoken_addrs_len, res);
+    //return (lptoken_addrs_len, array);
+}
+
+func map_address_to_poolinfo{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr} (
+    lpt_addrs: Address*,
+    poolinfo: PoolInfo*,
+    array_len: felt,
+    curr_index: felt
+) -> () {
+    let val = get_poolinfo(lpt_addrs[curr_index]);
+    assert poolinfo[curr_index] = val;
+    if(array_len == curr_index + 1){
+        return ();
+    }
+    return map_address_to_poolinfo(lpt_addrs, poolinfo, array_len, curr_index + 1);
+}
+
+// Returns UserPoolInfo, which is the value of user's capital in pool and PoolInfo.
+func get_userpoolinfo{
+    syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr
+}(
+    user_address: Address,
+    lptoken_address: Address
+) -> UserPoolInfo {
+    alloc_locals;
+    let poolinfo = get_poolinfo(lptoken_address);
+    let (balance: Uint256) = ILPToken.balanceOf(contract_address=lptoken_address, account=user_address);
+    if (balance.low == 0 and balance.high == 0){
+        let nonzero_val = Uint256(1, 0);
+        let res = UserPoolInfo(nonzero_val, poolinfo);
+        return res;
+    }
+
+    let (currency_address) = get_underlying_token_address(lptoken_address); // This is unnecessarry, right?
```

Comment:
underlying_token_address is equivalent to base_token_address, right? Then we save one piece of info in two different places if I am reading the spaghetti correctly.

Reply:
no... it depends on the pool
- call pool: underlying_token_address=ETH
- put pool: underlying_token_address=USD